src/avm2/native.ts

```diff
@@ -1369,71 +1372,127 @@ module Shumway.AVM2.AS {
+        p = p.replace(/\((?!\?)|\(\?P<([\w$]+)>/g, function (match) {
+            var name = match[1];
+            if (match[1]) {
+              if (name !== 'length') {
+                captureNames.push(name);
+              }
+              if (captureNames.indexOf(match[1]) > -1) {
+                // TODO: Handle the case were same name is used for multiple groups.
+              }
+            } else {
+              captureNames.push(null);
+            }
+            return '(';
+        });
+        var r = this._jsRegExp = new RegExp(p, f);
+        this.global = r.global;
+        this.multiline = r.multiline;
+        this.ignoreCase = r.ignoreCase;
+        this._source = pattern;
+    }
 
     get native_source(): string {
-      var self: any = this;
-      return self.source;
+      return this._source;
     }
 
     get native_global(): boolean {
-      var self: any = this;
-      return self.global;
+      return this.global;
     }
 
     get native_ignoreCase(): boolean {
-      var self: any = this;
-      return self.ignoreCase;
+      return this.ignoreCase;
     }
 
     get native_multiline(): boolean {
-      var self: any = this;
-      return self.multiline;
+      return this.multiline;
     }
 
     get native_lastIndex(): number /*int*/ {
-      var self: any = this;
-      return self.lastIndex;
+      return this._jsRegExp.lastIndex;
     }
 
     set native_lastIndex(i: number /*int*/) {
-      var self: any = this;
-      i = i | 0;
-      self.lastIndex = i;
+      this._jsRegExp.lastIndex = i | 0;
     }
 
     get native_dotall(): boolean {
-      var self: any = this;
-      return self.dotall;
+      return this.dotall;
     }
 
     get native_extended(): boolean {
-      var self: any = this;
-      return self.extended;
+      return this.extended;
     }
 
     exec(s: string = ""): any {
-      var result = RegExp.prototype.exec.apply(this, arguments);
+      var result = this._jsRegExp.exec(s);
+
       if (!result) {
         return result;
       }
-      // For some reason named groups in AS3 are set to the empty string instead of
-      // undefined as is the case for indexed groups. Here we just emulate the AS3
-      // behaviour.
-      var keys = Object.keys(result);
-      for (var i = 0; i < keys.length; i++) {
-        var k = keys[i];
-        if (!isNumeric(k)) {
-          if (result[k] === undefined) {
-            result[k] = "";
+
+      var captureNames = this._captureNames;
+      for (var i = 0; i < captureNames.length; i++) {
+          var name = captureNames[i];
+          if (name !== null) {
+              result[name] = result[0] || '';
+          } else {
+              // Groups in AS3 are set to the empty string instead of undefined.
+              result[name] = '';
           }
-        }
       }
+
       Shumway.AVM2.Runtime.publicizeProperties(result);
       return result;
     }
+
+    search(s: string): number {
+      return s.search(this._jsRegExp);
```

Comment:
Use `s.$Bgsearch` here, or `ASString.search.call(this._jsRegExp)`. Otherwise, you'll get the JS version of `String.prototype.search`.